i-VRESSE/workflow-builder
# Make array and table fields indexable

Fixes #57 
Refs #43

By adding `ui:indexable:true` to uiSchema you can have index in an array sub form and table sub form.

TODO:
* [x] add index to table field
* [x] add index to array field
* [x] generate catalog index with indexable:true in uiSchema 
    * for schema props with `maxItemsFrom: molecules` config
    * molecules parameter
    * ~for schema props wtih `type:'array'`~  dont care about order here so no index is needed either.

Out of scope
* use molecule file name as index (#76)

To test
1. In haddock3 download app,
    1.  `yarn dev`
    2. goto http://localhost:3000
    3. configure global parameters
        1. upload molecules
        2. should have a index (0, 1) in front each file name
    5. add mdref node from refinement category
        1. configure node
        2. expand molecule sub form
        3. Add several mol_shape rows
        4. In front of each row there should be an index.
    7. the haddock3 apps do not have a indexable table because all tables in the haddock3 catalogs do not care about order 
5. In kitchen sink app,
    1.  `yarn dev`
    3. goto http://localhost:3002
    6. add node in indexable category and configure
    7. both table and array node should have 0,1,2,3 after several rows have been added
4. In storybook of form package
    1. `yarn storybook`
    2. goto http://localhost:6003
    3. Open stories in Array folder and the last story in TableField folder.

as array it looks like
![localhost_3002_ (1)](https://user-images.githubusercontent.com/1713488/180437054-79f18670-0a27-48f4-99a3-0bedd06b7c7d.png)
and as table
![localhost_3002_ (2)](https://user-images.githubusercontent.com/1713488/180437105-79f319cf-c723-4e5c-af3a-e85243900eba.png)

Due to changes in haddock3 this PR also contains the global preprocess parameter.
